controllers/cloud.redhat.com/clowdenvironment_reconciliation.go

```diff
@@ -0,0 +1,489 @@
+
+			controllerutil.RemoveFinalizer(r.env, envFinalizer)
+			removeFinalizeErr := r.client.Update(*r.ctx, r.env)
+			if removeFinalizeErr != nil {
+				r.log.Info("Cloud not remove finalizer", "err", removeFinalizeErr)
+				return ctrl.Result{}, removeFinalizeErr
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+	return ctrl.Result{}, nil
+}
+
+//Perform finalization of the environment. Called by markedForDeleteion
+//Note: _ at beginning of method name indicates that a method is called
+//by a step method, not directly by the reconcilation loop
+func (r *ClowdEnvironmentReconciliation) _finalizeEnvironment() error {
+
+	provider := r._makeProvider()
+
+	err := runProvidersForEnvFinalize(*r.log, provider)
+	if err != nil {
+		return err
+	}
+
+	if r.env.Spec.TargetNamespace == "" {
+		namespace := &core.Namespace{}
+		namespace.SetName(r.env.Status.TargetNamespace)
+		r.log.Info(fmt.Sprintf("Removing auto-generated namespace for %s", r.env.Name))
+		r.recorder.Eventf(r.env, "Warning", "NamespaceDeletion", "Clowder Environment [%s] had no targetNamespace, deleting generated namespace", r.env.Name)
+		r.client.Delete(context.TODO(), namespace)
+	}
+	delete(managedEnvironments, r.env.Name)
+	managedEnvsMetric.Set(float64(len(managedEnvironments)))
+
+	delete(presentEnvironments, r.env.Name)
+	presentEnvsMetric.Set(float64(len(presentEnvironments)))
+
+	r.log.Info("Successfully finalized ClowdEnvironment")
+	return nil
+}
+
+//Adds a finalizer to the environment if one is not already present
+func (r *ClowdEnvironmentReconciliation) addFinalizerIfRequired() (ctrl.Result, error) {
+	// Add finalizer for this CR
+	if !contains(r.env.GetFinalizers(), envFinalizer) {
+		if addFinalizeErr := r._addFinalizer(); addFinalizeErr != nil {
+			r.log.Info("Cloud not add finalizer", "err", addFinalizeErr)
+			return ctrl.Result{}, addFinalizeErr
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+//Implementation method for addining a finalizer
+func (r *ClowdEnvironmentReconciliation) _addFinalizer() error {
+	r.log.Info("Adding Finalizer for the ClowdEnvironment")
+	controllerutil.AddFinalizer(r.env, envFinalizer)
+
+	// Update CR
+	err := r.client.Update(context.TODO(), r.env)
+	if err != nil {
+		r.log.Error(err, "Failed to update ClowdEnvironment with finalizer")
+		return err
+	}
+	return nil
+}
+
+//Request per provider methods if the config calls for it
+func (r *ClowdEnvironmentReconciliation) perProviderMetrics() (ctrl.Result, error) {
+	if clowderconfig.LoadedConfig.Features.PerProviderMetrics {
+		requestMetrics.With(prometheus.Labels{"type": "env", "name": r.env.Name}).Inc()
+	}
+	return ctrl.Result{}, nil
+}
+
+//Get or create and then update the cache for the target namespace
+func (r *ClowdEnvironmentReconciliation) initTargetNamespace() (ctrl.Result, error) {
+	result := ctrl.Result{}
+	var err error = nil
+
+	if r.env.Status.TargetNamespace != "" {
+		return result, err
+	}
+
+	if r.env.Spec.TargetNamespace == "" {
+		result, err = r._makeTargetNamespace()
```

Comment:
Why the underscores?

Reply:
Yeah, I wasn't sure how you'd react to that. I can change it if you don't like it and I'm not married to it. Basically, I wanted some way to denote what is and is not a step method. I could preface every step method with `step` but that would get ugly fast, and considering steps are 90% of the methods that didn't seem sensible. Scope can't help us because all these methods are private. If each step was its own type then each step's Run method would be public and any helper methods would be private and limited to that type. So, the best worst solution I landed on was to use an underscore to hint that this method is a private implementation detail and not a step.

Reply:
Renamed the underscore methods to funcNameImplementation in cases where funcName already existed as a step. Otherwise just gave them sane names.